pkg/controller/controllerutil/controllerutil.go

```diff
@@ -348,27 +348,31 @@ func mutate(f MutateFn, key client.ObjectKey, obj client.Object) error {
 // MutateFn is a function which mutates the existing object into it's desired state.
 type MutateFn func() error
 
-// AddFinalizer accepts an Object and adds the provided finalizer if not present.
-func AddFinalizer(o client.Object, finalizer string) {
+// AddFinalizer accepts an Object and adds the provided finalizer and returns true if not present.
+func AddFinalizer(o client.Object, finalizer string) bool {
 	f := o.GetFinalizers()
 	for _, e := range f {
 		if e == finalizer {
-			return
+			return false
```

Comment:
This addition does seem useful. But the other way which I could think of it is this:
1. Case 1: Finalizer is present and user wants it to be or finalizer is not present and user wants it removed - In this case, we can just use `ConatinsFinalizer()`, which is a linear operation of complexity O(n).
2. Case 2: When the finalizer is present and user wants it removed or vice versa: Here there needs to be an update to the object which needs to happen. As you have explained, we would have to use `ContainsFinalizer` + `Add/RemoveFinalizer` where there would be linear iteration twice. The complexity still remains linear. I am not sure if there would be much performance improvement by saving on one iteration, since updating the object is anyway a necessity. 

We could still return a signal in Add or Remove Finalizer saying it has been updated or not, but I think boolean may not be apt. Returning `false` in `AddFinalizer` can also be interpreted by users that it has not been added to the object at all (though with proper docs we could explain that its just adding element to a list). We can instead maybe return an indication that is already present. 

```go
type FinalizerStatus string

const (
	FinalizerUpdated    FinalizerStatus = "updated"
	FinalizerNotUpdated FinalizerStatus = "notupdated"
)

// AddFinalizer accepts an Object and adds the provided finalizer if not present.
func AddFinalizer(o client.Object, finalizer string) FinalizerStatus {
	f := o.GetFinalizers()
	for _, e := range f {
		if e == finalizer {
			return FinalizerUpdated
		}
	}
	o.SetFinalizers(append(f, finalizer))
	return FinalizerNotUpdated
}
```

However, this would be a breaking change. Also, I think the [finalizer](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/finalizer/finalizer.go) library already implements this in a better way.

Reply:
Thanks for the fast response!

The finalizer library contains one of the issues I intend to solve, namely two linear iterations, e.g. for Add:
https://github.com/kubernetes-sigs/controller-runtime/blob/4e7f0c968b90a306fb9ef22f8e92fa36b98b4ef1/pkg/finalizer/finalizer.go#L57
https://github.com/kubernetes-sigs/controller-runtime/blob/4e7f0c968b90a306fb9ef22f8e92fa36b98b4ef1/pkg/finalizer/finalizer.go#L58
It does provide a bool indicating whether the object was updated: https://github.com/kubernetes-sigs/controller-runtime/blob/4e7f0c968b90a306fb9ef22f8e92fa36b98b4ef1/pkg/finalizer/finalizer.go#L59

The functions are called `AddFinalizer` and `RemoveFinalizer` so I think it is intuitive and logical to interpret a boolean return code as an indication of whether they did what their names indicate they should.

I would be ok with wrapping the boolean in a struct just to give it a name that the programmer must type, like the library does, but its struct has more than one field.

The string solution takes more work to match.

I propose naming the bool return value `finalizersUpdated` in the function signature to avoid misinterpretation.  I updated the pull request with that solution.